src/corefx/System.Security.Cryptography.Native/openssl.c

```diff
@@ -0,0 +1,625 @@
+    {
+        return x509->sig_alg->algorithm;
+    }
+
+    return NULL;
+}
+
+/*
+Function:
+GetX509PublicKeyParameterBytes
+
+Used by System.Security.Cryptography.X509Certificates' OpenSslX509CertificateReader to copy out the
+parameters to the algorithm used by the certificate public key
+
+Return values:
+0: Invalid X509 pointer
+1: Data was copied
+Any negative value: The input buffer size was reported as insufficient. A buffer of size ABS(return) is required.
+*/
+int
+GetX509PublicKeyParameterBytes(
+    X509* x509,
+    unsigned char* pBuf,
+    int cBuf)
+{
+    if (!x509 || !x509->cert_info || !x509->cert_info->key || !x509->cert_info->key->algor)
+    {
+        return 0;
+    }
+
+    ASN1_TYPE* parameter = x509->cert_info->key->algor->parameter;
+    int len = i2d_ASN1_TYPE(parameter, NULL);
+
+    if (cBuf < len)
+    {
+        return -len;
+    }
+
+    unsigned char* pBuf2 = pBuf;
+    len = i2d_ASN1_TYPE(parameter, &pBuf2);
+
+    if (len > 0)
+    {
+        return 1;
+    }
+
+    return 0;
+}
+
+/*
+Function:
+GetX509PublicKeyBytes
+
+Used by System.Security.Cryptography.X509Certificates' OpenSslX509CertificateReader to obtain the
+raw bytes of the public key.
+
+Return values:
+NULL if the public key cannot be determined, a pointer to the ASN1_BIT_STRING structure representing
+the public key.
+*/
+ASN1_BIT_STRING*
+GetX509PublicKeyBytes(
+    X509* x509)
+{
+    if (x509 && x509->cert_info && x509->cert_info->key)
+    {
+        return x509->cert_info->key->public_key;
+    }
+
+    return NULL;
+}
+
+/*
+Function:
+GetAsn1StringBytes
+
+Used by the NativeCrypto shim type to extract byte[] data from OpenSSL ASN1_* types whenever a byte[] is called
+for in managed code.
+
+Return values:
+0: Invalid X509 pointer
+1: Data was copied
+Any negative value: The input buffer size was reported as insufficient. A buffer of size ABS(return) is required.
+
+Remarks:
+ Many ASN1 types are actually the same type in OpenSSL:
+   STRING
+   INTEGER
+   ENUMERATED
+   BIT_STRING
+   OCTET_STRING
+   PRINTABLESTRING
+   T61STRING
+   IA5STRING
+   GENERALSTRING
+   UNIVERSALSTRING
+   BMPSTRING
+   UTCTIME
+   TIME
+   GENERALIZEDTIME
+   VISIBLEStRING
+   UTF8STRING
+
+ So this function will really work on all of them.
+*/
+int
+GetAsn1StringBytes(
+    ASN1_STRING* asn1,
+    unsigned char* pBuf,
+    int cBuf)
+{
+    if (!asn1)
+    {
+        return 0;
+    }
+
+    if (!pBuf || cBuf < asn1->length)
+    {
+        return -asn1->length;
+    }
+
+    memcpy(pBuf, asn1->data, asn1->length);
+    return 1;
+}
+
+/*
+Function:
+GetX509NameRawBytes
+
+Used by System.Security.Cryptography.X509Certificates' OpenSslX509CertificateReader to obtain the
+DER encoded value of an X500DistinguishedName.
+
+Return values:
+0: Invalid X509 pointer
+1: Data was copied
+Any negative value: The input buffer size was reported as insufficient. A buffer of size ABS(return) is required.
+*/
+int
+GetX509NameRawBytes(
+    X509_NAME* x509Name,
```

Comment:
GetX509NameRawBytes and GetX509PublicKeyParameterBytes both return DER-encoded data, so there's no zero-length valid value (an empty item still has a tag byte and a length(0) byte).

GetAsn1StringBytes is the only thing that looks vulnerable, since it would return -0 if pBuf is NULL, but then will return 1 if pBuf wasn't NULL.  That one is merely a little silly.

I don't see any cases where we would want to see the error and throw, so I don't think anything has slipped through on the corners .